### PR TITLE
[JENKINS-42591] && [JENKINS-47060] Add overwrite option to zip step.

### PR DESCRIPTION
**Pull request by fabiodcasilva**
Added the possibility to overwrite the zip file in cases where the destination file already exists.

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStepExecution.java [hunk 1]
```diff
@@ -120,23 +120,30 @@ protected Void run() throws Exception {
...
             this.glob = StringUtils.isBlank(glob) ? "**/*" : glob;
+            this.overwrite = overwrite;
         }
 
         @Override
         public Integer invoke(File dir, VirtualChannel channel) throws IOException, InterruptedException {
-            String canonicalZip = new File(zipFile.getRemote()).getCanonicalPath();
+            File zip = new File(zipFile.getRemote());
+            if (overwrite && zip.exists()) {
+                if (!zip.delete()) {
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStepExecution.java:135]: https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#delete-java.nio.file.Path- would be a better File management API

File: src/main/resources/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStep/help-overwrite.html [hunk 2]
```diff
@@ -0,0 +1,27 @@
+<!--
+  ~ The MIT License (MIT)
+  ~
+  ~ Copyright (c) 2016 CloudBees Inc.
```
[the remark this answers is not in the document]
Comment (fabiodcasilva) [reply src/main/resources/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStep/help-overwrite.html]: I updated it. Thanks

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStepExecution.java (hunk 1 above)
Comment (fabiodcasilva) [reply src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStepExecution.java]: I changed it to the api you suggested. 
@oleg-nenashev can you check if it is as expected?
Thank you very much

File: src/main/resources/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStep/config.groovy [hunk 3]
```diff
@@ -36,8 +36,11 @@ f.entry(field: 'dir', title: _('Directory')) {
...
 }
+
 f.entry(field: 'archive', title: _('Archive as artifact')) {
     f.checkbox()
 }
 
-
+f.entry(field: 'overwrite', title: _('Overwrite')) {
+    f.checkbox()
+}
```
Comment (rsandell) [inline src/main/resources/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStep/config.groovy:46]: Just a reminder to myself that all the optional arguments should be put behind an advanced section.

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStepExecution.java [hunk 4]
```diff
@@ -120,15 +122,20 @@ protected Void run() throws Exception {
...
             this.zipFile = zipFile;
             this.glob = StringUtils.isBlank(glob) ? "**/*" : glob;
+            this.overwrite = overwrite;
         }
 
         @Override
         public Integer invoke(File dir, VirtualChannel channel) throws IOException, InterruptedException {
-            String canonicalZip = new File(zipFile.getRemote()).getCanonicalPath();
+            String canonicalZip = zipFile.getRemote();
+            if (overwrite && !Files.deleteIfExists(Paths.get(canonicalZip))) {
```
Comment (deafol) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStepExecution.java:136]: So now it throws the exception (and fails) if the zip does not exist and overwrite=true. Was that intended?

My case: I need the file to be overwritten when it exists + be created when it does not. That used to work without specifying the overwrite flag (which wasn't there obviously)

Comment (rsandell) [reply src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStepExecution.java:136]: No, that sounds a bit weird indeed. :thinking:

Comment (espguitarist33) [reply src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStepExecution.java:136]: This seems like a bug

Comment (clamprecht01) [reply src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStepExecution.java:136]: I am thinking the only workaround for now is to check if file exists before attempting to zip.

**Pull request merged**